monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/MergePrioritizedListObservable.scala

```diff
@@ -0,0 +1,174 @@
+  * observable that eagerly emits source items downstream as soon as demand is
+  * signaled, choosing the item from the highest priority (greater numbers
+  * mean higher priority) source when items from multiple sources are
+  * available. If items are available from multiple sources with the same
+  * highest priority, one of them is chosen arbitrarily.
+  *
+  * Source items are buffered only to the extent necessary to accommodate
+  * backpressure from downstream, and thus if only a single item is available
+  * when demand is signaled, it will be emitted regardless of priority.
+  *
+  * Backpressure is propagated from downstream to the source observables, so
+  * that items from a given source will always be emitted downstream in the
+  * same order as received from the source, and at most a single item from a
+  * given source will be in flight at a time.
+  */
+private[reactive] final class MergePrioritizedListObservable[A](sources: Seq[(Int, Observable[A])])
+  extends Observable[A] {
+
+  override def unsafeSubscribeFn(out: Subscriber[A]): Cancelable = {
+    import out.scheduler
+
+    val numberOfObservables = sources.size
+
+    val lock = new AnyRef
+    var isDone = false
+
+    // MUST BE synchronized by `lock`
+    var lastAck = Continue: Future[Ack]
+
+    case class PQElem(data: A, promise: Promise[Ack], priority: Int)
+
+    object PQElemOrdering extends Ordering[PQElem] {
+      override def compare(x: PQElem, y: PQElem): Int =
+        x.priority.compareTo(y.priority)
+    }
+
+    val pq = new mutable.PriorityQueue[PQElem]()(PQElemOrdering)
+
+    // MUST BE synchronized by `lock`
+    var completedCount = 0
+
+    // MUST BE synchronized by `lock`
+    def rawOnNext(a: A): Future[Ack] = {
+      if (isDone) Stop else out.onNext(a)
+    }
+
+    // MUST BE synchronized by `lock`
+    def processNext(): Future[Ack] = {
+      val e = pq.dequeue()
+      val fut = rawOnNext(e.data)
+      e.promise.completeWith(fut)
+      fut
+    }
+
+    // MUST BE synchronized by `lock`
+    def signalOnNext(): Future[Ack] = {
+      lastAck = lastAck match {
+        case Continue => processNext()
+        case Stop => Stop
+        case async =>
+          async.flatMap {
+            // async execution, we have to re-sync
+            case Continue => lock.synchronized(processNext())
+            case Stop => Stop
+          }
+      }
+      lastAck
+    }
+
+    def signalOnError(ex: Throwable): Unit =
+      lock.synchronized {
+        if (!isDone) {
+          isDone = true
+          out.onError(ex)
+          lastAck = Stop
+          completePromises()
+        }
+      }
+
+    def signalOnComplete(): Unit =
+      lock.synchronized {
+        completedCount += 1
+
+        if (completedCount == numberOfObservables && !isDone) {
+          lastAck match {
+            case Continue =>
+              isDone = true
+              out.onComplete()
+              completePromises()
```

Comment:
Regarding the issue:
> The key is that so long as there are items the downstream hasn't yet processed, lastAck will remain a Future[Ack], and signalOnComplete won't call out.onComplete until that future completes, i.e., downstream has finished processing all pending items including the last one from the last source.
>
>Take the simplest case of your example: 2 sources w/ 1 item each:
>
>Source A calls onNext, item gets sent downstream: lastAck = Future[Ack for A's item]
>Source A calls onComplete: completedCount = 1
>Source B calls onNext, item gets queued: lastAck = Future[Ack for A's item].flatMap(Future[Ack for B's item])
>Source B calls onComplete: completedCount = 2; lastAck.onComplete { ... out.onComplete }

Looking at the contract: https://monix.io/docs/3x/reactive/observers.html#contract
> 3. Back-pressure for onComplete and onError is optional: when calling onComplete or onError you are not required to wait on the Future[Ack] of the previous onNext.

It implies that Source B could call:

```scala
out.onNext(lastElem)
out.onComplete()
```
 
In that case, we could send remaining elements from the queue to the downstream instead of completing each promise with `Stop`.

Reply:
> In that case, we could send remaining elements from the queue to the downstream instead of completing each promise with `Stop`.

So long as the downstream returns `Continue` and none of the upstreams call `onError`, the downstream will always get all the items emitted by the upstreams. That's what the "should push all items downstream before calling onComplete" test is verifying. Again:

> The key is that so long as there are items the downstream hasn't yet processed, lastAck will remain a Future[Ack], and signalOnComplete won't call out.onComplete (or completePromises) until that future completes, i.e., downstream has finished processing all pending items including the last one from the last source.